### PR TITLE
feat(Host): RHICOMPL-1397 add has_policy attribute

### PR DESCRIPTION
**Pull request by akofink**
Since we added a search term, it makes sense to also add an attribute so users can know what to expect when using the search.

Signed-off-by: Andrew Kofink <akofink@redhat.com>

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [X] Input Validation
- [X] Output Encoding
- [X] Authentication and Password Management
- [X] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: app/models/host.rb [hunk 1]
```diff
@@ -25,6 +25,11 @@ def readonly?
   alias destroy save
   alias delete save
 
+  def policy_hosts?
+    policy_hosts.any?
```
Comment (vkrizan) [inline app/models/host.rb:29]: Observation: Preload policy hosts?

**Review by vkrizan — COMMENTED** [review]
Code looks good so far. Will test in a bit...

Personally, I don't like adding single-purpose flags to API, especially when the information can be retrieved by other/general means. What is the difference between `has_policy` and checking profiles (relationship)?

**Comment by akofink** [comment]
Unfortunately, the profiles relationship on a host is `all_profiles`, which includes test result profiles as well as policy profiles. We'd need to add another relationship (or two), like `test_result_profiles` and `policy_profiles`.

File: app/controllers/concerns/rendering.rb [hunk 2]
```diff
@@ -6,9 +6,8 @@ module Rendering
 
   included do
     def render_json(model, **args)
-      opts = ['index'].include?(action_name) ? metadata : {}
-      opts.merge!(include: params[:include].split(',')) if params[:include]
-      render({ json: serializer.new(model, opts) }.merge(args))
+      model = model.includes(includes) if index? && includes
+      render({ json: serializer.new(model, serializer_opts) }.merge(args))
```
Comment (akofink) [inline app/controllers/concerns/rendering.rb:10]: Rubocop forced a refactor of this method; it was getting pretty complex.

File: app/controllers/concerns/rendering.rb [hunk 3]
```diff
@@ -38,5 +48,7 @@ def model_errors(models = [])
     def serializer
       raise NotImplementedError
     end
+
+    def includes; end
```
Comment (akofink) [inline app/controllers/concerns/rendering.rb:52]: @vkrizan I've added a new feature to this rendering concern which allows controllers to specify any includes to perform prior to serialization.

File: app/controllers/v1/systems_controller.rb [hunk 4]
```diff
@@ -25,5 +25,9 @@ def serializer
     def resource
       Host
     end
+
+    def includes
+      %i[policy_hosts]
+    end
```
Comment (akofink) [inline app/controllers/v1/systems_controller.rb:31]: And here is the systems_controller using the new includes rendering feature.

**Comment by akofink** [comment]
Ok, @vkrizan, please let me know what you think. I've opted to add includes as a part of our rendering code so we can more easily correct N+1 issues in the future. I've only included `policy_hosts` to the hosts API for now, but we can revisit this and other controllers to easily add more `includes`.

File: app/controllers/concerns/rendering.rb (hunk 3 above)
Comment (vkrizan) [reply app/controllers/concerns/rendering.rb:52]: :+1:

**Review by vkrizan — APPROVED** [review]
Works as expected.

> Unfortunately, the profiles relationship on a host is all_profiles, which includes test result profiles as well as policy profiles. We'd need to add another relationship (or two), like test_result_profiles and policy_profiles.

There is nothing stopping us to add those relationships to REST API. We already have them in the GraphQL.

**Pull request merged**